File: services/agent/attachments.py
"""Attachments on a chat turn: what is accepted, and what it costs.

One turn may carry a few files. Two kinds reach the model and they reach it by
completely different routes, which is the reason this module exists rather than
a helper in the blueprint:

* **An image** goes to the provider as an image part, built here into agno's
  ``Image`` and passed to ``agent.run(images=...)``. Agno's LiteLLM model turns
  each one into ``{"type": "image_url", "image_url": {"url": "data:...;base64,..."}}``
  through ``agno.utils.openai.images_to_message``, so the bytes are base64 on
  the wire twice: once from the browser to us, once from us to the provider.
* **A text file** never becomes a media part at all. It is decoded and folded
  into the message inside an ``<attachment>`` block, because a text file is
  data somebody wrote and every other piece of text that entered the system as
  data goes through the same wrapper.

Three rules this module exists to keep
--------------------------------------

**A model that cannot see refuses the turn, by name.** Not here: this module
reports that an image is present and :func:`services.agent.builder.build_agent`
refuses. What matters is that no path silently drops an image. An operator who
attaches a screenshot and gets a confident answer has no way to tell it was
never looked at, and that is the outcome worth engineering against.

**The bytes decide what a file is.** The declared media type and the filename
are both attacker-influenced, and neither is consulted for the allow decision.
A file is an image because it starts with a PNG, JPEG, GIF or WebP signature,
and it is text because every byte of it decodes as UTF-8 with no control
characters in it. A declared type that contradicts the bytes is refused rather
than reinterpreted: the client is either confused or lying, and quietly picking
the answer we prefer is how a mismatch stops being visible.

**Everything is bounded.** An image is base64 in a JSON body, so it inflates by
about a third on the wire, and then it is charged as input tokens **again on
every later turn of the conversation**, because agno replays the history. The
cost of one image is therefore not paid once. Measured against the configured
model: a 360x130 PNG cost 73 input tokens on its own turn and 70 more on the
next one; a 1920x1080 screenshot cost 2,449 and then 2,447, and goes on costing
that on every turn inside the ``DEFAULT_NUM_HISTORY_RUNS`` window. Note that
neither number tracks the byte count, because a provider charges an image by its
dimensions; the caps below bound bytes because bytes are what this process has
to hold, and the token cost is what the operator has to be told.

Storage
-------
:func:`stored_metadata` is what goes on the ``ag_message`` row: name, sniffed
type, byte count and a short digest. **The bytes are not stored.**

The reference is ``MAX_STORED_VIZ`` in ``blueprints/agent.py``, which caps a
turn at four chart specs on the grounds that "a chart spec is a series, not a
sentence" and an unbounded count would put a data dump in a text column. An
image is a far larger object than a chart spec and would go in the same column,
which is read in full every time the conversation is opened. Measured on this
change: a whole stored user row carrying two attachments is 467 bytes, of which
the metadata is about 120 bytes per file; the image it describes was 2,584
bytes, a 1920x1080 screenshot was 11,140, and the per-file cap is 4,000,000.
The metadata is enough to re-render the composer's chip, which is what the
transcript needs to show: that a file was sent, which one, and how big it was.
"""

from __future__ import annotations

import base64
import binascii
import hashlib
import re
from collections.abc import Sequence
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from services.agent.prompts import wrap_attachment
from utils.logging import get_logger

if TYPE_CHECKING:  # pragma: no cover - typing only, never imported at runtime
    from agno.media import Image

logger = get_logger(__name__)

__all__ = [
    "MAX_ATTACHMENTS",
    "MAX_ATTACHMENT_BYTES",
    "MAX_ENCODED_CHARS",
    "MAX_NAME_CHARS",
    "MAX_TEXT_CHARS",
    "MAX_TOTAL_BYTES",
    "Attachment",
    "AttachmentError",
    "has_image",
    "images_for_run",
    "parse_attachments",
    "prompt_block",
    "stored_metadata",
]

# ---------------------------------------------------------------------------
# The caps
#
# Every one of them is enforced before the next byte is decoded, and each says
# what it is protecting. None of them is configurable: this module takes no
# configuration from `.env`, and a limit an operator can raise is a limit that
# gets raised the first time a file does not fit.
# ---------------------------------------------------------------------------

#: Files per turn. A question about a screenshot needs one; a comparison needs
#: two or three. Beyond that the conversation is carrying a directory, and every
#: one of them is replayed into context on every subsequent turn.
MAX_ATTACHMENTS = 4

#: Decoded bytes in one file. A phone screenshot is 200 KB to 2 MB and a full
#: desktop capture is under 4 MB, so this accepts what an operator will actually
#: attach without accepting a video frame dump.
MAX_ATTACHMENT_BYTES = 4_000_000

#: Decoded bytes across the whole turn, which is the number that matters: four
#: files at the per-file cap would otherwise be 16 MB in one request body.
MAX_TOTAL_BYTES = 8_000_000

#: Characters of a text file that reach the prompt. Roughly 5,000 tokens, which
#: is a large fraction of a turn on its own, and the file is replayed on every
#: later turn. Longer files are refused rather than silently truncated: an
#: answer written from the first fifth of a file, with no sign that the rest was
#: dropped, is the same failure as an image that was never looked at.
MAX_TEXT_CHARS = 20_000

#: Characters of a filename that are kept. The name is an attribute in the
#: prompt and a label in the transcript, never a path and never opened.
MAX_NAME_CHARS = 120

#: Ceiling on the base64 string, checked before decoding so a 100 MB payload is
#: refused without being materialised a second time. Base64 is 4 characters per
#: 3 bytes; the slack covers padding, a data-URL prefix and any embedded
#: whitespace.
MAX_ENCODED_CHARS = (MAX_ATTACHMENT_BYTES * 4) // 3 + 4096

# ---------------------------------------------------------------------------
# Sniffing
# ---------------------------------------------------------------------------

#: Magic-byte prefixes for the image types that are accepted. Prefix matching
#: only: nothing here decodes an attacker-supplied image, because a decoder is a
#: far larger attack surface than a byte comparison and no decision needs the
#: pixels.
_IMAGE_SIGNATURES: tuple[tuple[bytes, str], ...] = (
    (b"\x89PNG\r\n\x1a\n", "image/png"),
    (b"\xff\xd8\xff", "image/jpeg"),
    (b"GIF87a", "image/gif"),
    (b"GIF89a", "image/gif"),
)

#: WebP is the one that is not a plain prefix: ``RIFF`` then a four byte size
#: then ``WEBP``.
_WEBP_HEAD = b"RIFF"
_WEBP_TAG = b"WEBP"

#: Formats that are definitely not text and are definitely not an image this
#: module accepts, named so the refusal says something useful. A PDF is the one
#: that needs this: its first bytes are plain ASCII, so without it a PDF would
#: be refused only once the reader reached the binary further in, and a short
#: one would be accepted as a text file whose entire content is ``%PDF-1.7``.
#: An operator will try a PDF, and "attach a text file instead" is a better
#: answer than the model summarising a version header.
_REJECTED_SIGNATURES: tuple[tuple[bytes, str], ...] = (
    (b"%PDF-", "a PDF"),
    (b"PK\x03\x04", "a zip or Office document"),
    (b"\x1f\x8b", "a gzip archive"),
    (b"Rar!\x1a\x07", "a RAR archive"),
    (b"7z\xbc\xaf\x27\x1c", "a 7-Zip archive"),
    (b"\x7fELF", "an executable"),
    (b"MZ", "an executable"),
    (b"%!PS", "a PostScript file"),
)

MIME_TEXT = "text/plain"

#: What "the client does not know what this is" looks like. An operating system
#: that cannot name a file says ``application/octet-stream``, which is not a
#: claim about the content and so cannot contradict one.
_UNDECLARED: frozenset[str] = frozenset({"", "application/octet-stream"})

#: Media types a client may declare for a file whose bytes sniff as text. A CSV,
#: a JSON export and a Markdown note are all text, and refusing them for their
#: label when their bytes are plainly text would be pedantry rather than safety.
_TEXT_DECLARATIONS: frozenset[str] = frozenset(
    {
        "text/plain",
        "text/csv",
        "text/markdown",
        "text/x-markdown",
        "text/tab-separated-values",
        "text/x-python",
        "text/xml",
        "text/html",
        "application/json",
        "application/x-ndjson",
        "application/xml",
        "application/csv",
    }
)

#: ``image/jpg`` is not a registered type but browsers and operating systems
#: emit it, and refusing a JPEG for spelling its own name the common way would
#: be a bug report rather than a defence.
_MIME_ALIASES: dict[str, str] = {"image/jpg": "image/jpeg", "image/pjpeg": "image/jpeg"}

#: Control characters that mean a file is not text. Tab, newline and carriage
#: return are excluded because a text file has them; everything else in C0, and
#: the C1 range, is a binary that happened to decode.
_NOT_TEXT = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\x9f]")

#: ``data:image/png;base64,`` and friends. The browser's ``readAsDataURL``
#: produces one, so accepting it saves every client the same split.
_DATA_URL = re.compile(r"\Adata:([A-Za-z0-9!#$&^_.+-]+/[A-Za-z0-9!#$&^_.+-]+)?;base64,", re.I)


class AttachmentError(ValueError):
    """An attachment was refused, with a message written for the operator.

    Attributes:
        message: Why it was refused, naming the file. Safe to return in an HTTP
            body: it never carries file content, only the file's own label and
            the measured facts about it.
    """

    def __init__(self, message: str) -> None:
        """Store the message on the exception as well as in its args."""
        super().__init__(message)
        self.message = message


@dataclass(frozen=True, slots=True)
class Attachment:
    """One accepted file.

    Frozen because an attachment crosses into the run thread with the rest of
    the turn and nothing may change it after it was measured and allowed.

    Attributes:
        name: The filename as the client sent it, trimmed to
            :data:`MAX_NAME_CHARS` and stripped of anything path-like. A label,
            never a path: nothing in this module opens a file.
        kind: ``image`` or ``text``, decided by the bytes.
        mime: The media type the bytes sniff as.
        declared: The media type the client declared, kept for the record and
            for the refusal message. It decides nothing.
        size: Decoded byte count.
        digest: First 12 hex characters of the SHA-256 of the bytes. Enough to
            recognise the same file attached twice, and to match a stored chip
            against a file on disk, without keeping the file.
        data: The image bytes. Empty for a text attachment.
        text: The decoded text. Empty for an image.
    """

    name: str
    kind: str
    mime: str
    declared: str
    size: int
    digest: str
    data: bytes = b""
    text: str = ""


def _clean_name(value: Any, index: int) -> str:
    """Reduce a client-supplied filename to a safe label.

    Args:
        value: Whatever the client sent as ``name``.
        index: Position in the list, used when there is no usable name.

    Returns:
        A single-line label of at most :data:`MAX_NAME_CHARS` characters. Path
        separators are dropped rather than escaped, because the value is only
        ever displayed and interpolated into a prompt attribute, and a name that
        still looks like a path invites a later reader to treat it as one.
    """
    text = "" if value is None else str(value)
    text = text.replace("\\", "/").rsplit("/", 1)[-1]
    text = _NOT_TEXT.sub("", text).replace("\n", " ").replace("\r", " ").strip()
    return text[:MAX_NAME_CHARS] or f"attachment-{index + 1}"


def _clean_mime(value: Any) -> str:
    """Normalise a declared media type.

    Args:
        value: Whatever the client sent as ``mime``, possibly with parameters.

    Returns:
        The lower-case type without parameters, aliases resolved, or an empty
        string when the client declared nothing.
    """
    text = "" if value is None else str(value)
    text = text.split(";", 1)[0].strip().lower()
    return _MIME_ALIASES.get(text, text)


def _sniff(data: bytes) -> tuple[str, str]:
    """Decide what a file is from its bytes alone.

    Args:
        data: The decoded file.

    Returns:
        ``(kind, mime)`` where kind is ``image`` or ``text``.

    Raises:
        AttachmentError: The bytes are neither an allowed image nor text.
    """
    for signature, mime in _IMAGE_SIGNATURES:
        if data.startswith(signature):
            return "image", mime
    if len(data) >= 12 and data[:4] == _WEBP_HEAD and data[8:12] == _WEBP_TAG:
        return "image", "image/webp"

    for signature, description in _REJECTED_SIGNATURES:
        if data.startswith(signature):
            raise AttachmentError(
                f"is {description}, which is not supported. Attach a PNG, JPEG, GIF or "
                "WebP image, or a text file."
            )

    try:
        text = data.decode("utf-8")
    except UnicodeDecodeError:
        raise AttachmentError(
            "is not a supported file. Attach a PNG, JPEG, GIF or WebP image, or a text file."
        ) from None
    if _NOT_TEXT.search(text):
        raise AttachmentError(
            "is not a supported file. Attach a PNG, JPEG, GIF or WebP image, or a text file."
        )
    return "text", MIME_TEXT


def _check_declaration(kind: str, sniffed: str, declared: str) -> None:
    """Refuse a declared media type that the bytes contradict.

    The bytes have already decided. This exists because a contradiction means
    the client is confused or lying about the file, and silently going with the
    bytes would make that invisible: a build that renames every upload to
    ``image/png`` would look like it worked right up until the day the rename
    was wrong about something that mattered.

    Args:
        kind: ``image`` or ``text``, from :func:`_sniff`.
        sniffed: The media type the bytes are.
        declared: The media type the client said, already normalised.

    Raises:
        AttachmentError: The declaration and the bytes disagree.
    """
    if declared in _UNDECLARED:
        return
    if kind == "image":
        if declared != sniffed:
            raise AttachmentError(
                f"was sent as {declared} but its bytes are {sniffed}. "
                "Send the file with its real type."
            )
        return
    if declared not in _TEXT_DECLARATIONS:
        raise AttachmentError(
            f"was sent as {declared} but its bytes are plain text. "
            "Send the file with its real type."
        )


def _decode(raw: Any) -> tuple[bytes, str]:
    """Decode one attachment's payload, refusing before it is materialised.

    Args:
        raw: The ``data`` field: base64, optionally as a ``data:`` URL.

    Returns:
        ``(bytes, mime_from_data_url)``. The second value is a declaration like
        any other and is only used when the client sent no ``mime`` field.

    Raises:
        AttachmentError: The payload is missing, too long, or not base64.
    """
    if not isinstance(raw, str) or not raw.strip():
        raise AttachmentError("carries no data.")

    payload = raw.strip()
    inline_mime = ""
    match = _DATA_URL.match(payload)
    if match:
        inline_mime = _clean_mime(match.group(1))
        payload = payload[match.end() :]

    # Before the decode, not after. Base64 is 4 characters per 3 bytes, so a
    # payload longer than this cannot fit under the byte cap and decoding it
    # would put a second copy of an oversized string in memory to learn that.
    if len(payload) > MAX_ENCODED_CHARS:
        raise AttachmentError(
            f"is larger than the {MAX_ATTACHMENT_BYTES // 1000} kB limit for one file."
        )

    try:
        data = base64.b64decode(payload, validate=True)
    except (binascii.Error, ValueError):
        raise AttachmentError("is not valid base64.") from None
    if not data:
        raise AttachmentError("is empty.")
    return data, inline_mime


def parse_attachments(raw: Any) -> list[Attachment]:
    """Validate a turn's attachments and return the accepted ones.

    Called before the conversation row is created, so a refused attachment
    leaves nothing behind.

    Args:
        raw: The request body's ``attachments`` field. ``None`` and an empty
            list both mean no attachments.

    Returns:
        The attachments, in the order they were sent.

    Raises:
        AttachmentError: Any of them was refused. The whole turn is refused
            rather than the offending file being dropped, because a turn that
            silently went out without the file the question was about is the
            failure this module exists to prevent.
    """
    if raw is None or raw == []:
        return []
    if not isinstance(raw, list):
        raise AttachmentError("attachments must be a list.")
    if len(raw) > MAX_ATTACHMENTS:
        raise AttachmentError(
            f"A turn may carry at most {MAX_ATTACHMENTS} attachments; {len(raw)} were sent."
        )

    accepted: list[Attachment] = []
    total = 0
    for index, item in enumerate(raw):
        if not isinstance(item, dict):
            raise AttachmentError(f"Attachment {index + 1} must be a JSON object.")
        name = _clean_name(item.get("name"), index)

        try:
            data, inline_mime = _decode(item.get("data"))
            if len(data) > MAX_ATTACHMENT_BYTES:
                raise AttachmentError(
                    f"is {len(data) // 1000} kB, over the "
                    f"{MAX_ATTACHMENT_BYTES // 1000} kB limit for one file."
                )
            kind, mime = _sniff(data)
            declared = _clean_mime(item.get("mime")) or inline_mime
            _check_declaration(kind, mime, declared)
        except AttachmentError as exc:
            # The file's own label is prefixed here rather than at every raise,
            # so each check states one fact and this states which file it is
            # about.
            raise AttachmentError(f"{name} {exc.message}") from None

        total += len(data)
        if total > MAX_TOTAL_BYTES:
            raise AttachmentError(
                f"The attachments total more than the {MAX_TOTAL_BYTES // 1000} kB "
                "limit for one turn."
            )

        text = ""
        if kind == "text":
            text = data.decode("utf-8")
            if len(text) > MAX_TEXT_CHARS:
                raise AttachmentError(
                    f"{name} is {len(text)} characters, over the {MAX_TEXT_CHARS} character "
                    "limit for a text attachment. Attach a shorter extract."
                )

        accepted.append(
            Attachment(
                name=name,
                kind=kind,
                mime=mime,
                declared=declared,
                size=len(data),
                digest=hashlib.sha256(data).hexdigest()[:12],
                data=data if kind == "image" else b"",
                text=text,
            )
        )

    logger.info(
        "Agent turn carries %s attachment(s): %s",
        len(accepted),
        ", ".join(f"{a.name} {a.mime} {a.size}B" for a in accepted),
    )
    return accepted


def has_image(attachments: Sequence[Attachment]) -> bool:
    """Whether the turn carries at least one image.

    Args:
        attachments: The accepted attachments.

    Returns:
        True when a vision-capable model is required for this turn.
    """
    return any(item.kind == "image" for item in attachments)


def images_for_run(attachments: Sequence[Attachment]) -> list[Image]:
    """Build agno's image objects for ``agent.run(images=...)``.

    Agno is imported here rather than at module level so this module stays
    importable without the optional dependency, which is the same reason
    ``tools/__init__.py`` imports it inside its factory.

    Args:
        attachments: The accepted attachments. Text ones are skipped.

    Returns:
        One ``agno.media.Image`` per image attachment, carrying raw bytes and
        the sniffed media type. ``content`` rather than ``url`` or ``filepath``:
        a URL would have agno fetch it server-side, and a path would mean this
        process wrote an operator's file to disk to read it straight back.
    """
    from agno.media import Image as AgnoImage

    return [
        AgnoImage(content=item.data, mime_type=item.mime, format=item.mime.split("/", 1)[-1])
        for item in attachments
        if item.kind == "image"
    ]


def prompt_block(attachments: Sequence[Attachment]) -> str:
    """Render the text attachments as untrusted blocks for the model.

    Args:
        attachments: The accepted attachments. Image ones are skipped, because
            an image travels as a media part rather than as prompt text.

    Returns:
        The blocks joined by blank lines, or an empty string when the turn
        carries no text file.
    """
    blocks = [
        wrap_attachment(item.name, item.text, media_type=item.mime, bytes=item.size)
        for item in attachments
        if item.kind == "text"
    ]
    return "\n\n".join(blocks)


def stored_metadata(attachments: Sequence[Attachment]) -> list[dict[str, Any]]:
    """What is persisted on the ``ag_message`` row.

    Args:
        attachments: The accepted attachments.

    Returns:
        One small mapping per file. No bytes and no text: see this module's own
        docstring for why, and for the size comparison that decided it.
    """
    return [
        {
            "name": item.name,
            "kind": item.kind,
            "mime": item.mime,
            "size": item.size,
            "digest": item.digest,
        }
        for item in attachments
    ]

File: frontend/src/lib/agent/attachments.ts
/**
 * Files on their way into a turn.
 *
 * The wire shape is `services/agent/attachments.py`: each file travels as
 * `{name, mime, data}` inside the JSON body of `POST /agent/api/chat/stream`,
 * where `data` is base64 and may carry its own `data:` URL prefix. That is
 * exactly what `FileReader.readAsDataURL` produces, so a browser needs no
 * re-encoding step and this module does none.
 *
 * **The server decides what a file is; this module only counts.** The bytes are
 * sniffed there, against magic-byte signatures, and a declared type that
 * contradicts them is refused by name. Repeating that check here would be a
 * second sniffer to keep in step with the first, so the only rules kept on this
 * side are the caps, which are the ones worth enforcing before a 4 MB upload
 * rather than after it. The numbers below are mirrored from that module and a
 * comment there is what has to change if they move.
 *
 * Two decisions that are not obvious:
 *
 * - **`mime` is forwarded only when the server has a rule for it.** `file.type`
 *   is the operating system guessing from an extension, and it guesses things
 *   like `application/x-yaml` that the server reads as a declaration
 *   contradicting plain text. An omitted type means "the client does not know",
 *   which cannot contradict anything, so the bytes decide and the operator
 *   never sees a refusal about a type they did not choose. A type the server
 *   does act on, `image/jpeg` on a file that is really a PNG, is still passed
 *   through, because that mismatch is one worth reporting.
 *
 * - **No object URL is ever created, so there is none to revoke.** The data URL
 *   is already in hand for the request, and `<img src={dataUrl}>` renders from
 *   the same string. `URL.createObjectURL` would allocate a second handle per
 *   thumbnail whose lifetime nothing in a chat thread naturally ends, which is
 *   a leak that survives until the tab closes.
 */

/**
 * Files one turn may carry.
 *
 * This and the two size caps below mirror `MAX_ATTACHMENTS`,
 * `MAX_ATTACHMENT_BYTES` and `MAX_TOTAL_BYTES` in
 * `services/agent/attachments.py`. The server is the authority; these exist so
 * a refusal happens before the bytes are read rather than after they are sent.
 */
export const MAX_ATTACHMENTS = 4

/** Decoded bytes in one file. */
export const MAX_ATTACHMENT_BYTES = 4_000_000

/** Decoded bytes across one turn. */
export const MAX_TOTAL_BYTES = 8_000_000

/** What a file is, for an icon and for the vision question. */
export type AttachmentKind = 'image' | 'text'

/** One file the operator has attached but not yet sent. */
export interface AgentAttachment {
  /** Stable for the life of the chip, so React keys and removal agree. */
  id: string
  name: string
  kind: AttachmentKind
  /** The type the browser declared, or empty when it declared nothing. */
  mime: string
  size: number
  /** `data:<type>;base64,<...>` exactly as FileReader produced it. */
  dataUrl: string
}

/**
 * What a sent turn remembers about a file.
 *
 * The same fields `services/agent/attachments.stored_metadata` writes onto the
 * message row, so a live turn and a reloaded one render the same chips. The
 * bytes are deliberately absent from both: an image is orders of magnitude
 * larger than the text column it would sit in, and it is read in full every
 * time the conversation is opened.
 */
export interface AgentAttachmentMeta {
  name: string
  kind: AttachmentKind
  mime: string
  size: number
}

/** The payload one file becomes in the request body. */
export interface AttachmentPayload {
  name: string
  mime?: string
  data: string
}

/**
 * Media types the server has a rule for.
 *
 * Anything else is dropped from the request rather than sent, for the reason
 * in the module docstring. Mirrors `_IMAGE_SIGNATURES` and
 * `_TEXT_DECLARATIONS` in `services/agent/attachments.py`.
 */
const DECLARABLE: ReadonlySet<string> = new Set([
  'image/png',
  'image/jpeg',
  'image/jpg',
  'image/pjpeg',
  'image/gif',
  'image/webp',
  'text/plain',
  'text/csv',
  'text/markdown',
  'text/x-markdown',
  'text/tab-separated-values',
  'text/x-python',
  'text/xml',
  'text/html',
  'application/json',
  'application/x-ndjson',
  'application/xml',
  'application/csv',
])

/**
 * What the file dialog offers first.
 *
 * A hint rather than a gate: every operating system treats it differently and
 * a file can always be dragged in past it, which is why the server sniffs.
 */
export const ATTACHMENT_ACCEPT =
  'image/png,image/jpeg,image/gif,image/webp,.txt,.csv,.tsv,.md,.json,.py,.xml,.html,.log'

/**
 * Base64 prefixes of the image signatures the server accepts.
 *
 * This chooses a thumbnail, never an allow: a file is attachable because the
 * caps allow it, and it is an image because the server said so after reading
 * its bytes. Reading it from the encoded string rather than from `file.type`
 * means a screenshot pasted with no declared type still shows itself.
 */
const IMAGE_BASE64_PREFIXES = ['iVBORw0KGgo', '/9j/', 'R0lGODdh', 'R0lGODlh', 'UklGR']

let attachmentSeq = 0

/**
 * A size an operator can read.
 *
 * Decimal, because the server's refusals are decimal: it reports the per-file
 * cap as "4000 kB". Dividing by 1024 here would print 3.8 MB for the same
 * limit, and a browser and a server disagreeing about how big a file is
 * allowed to be is how a cap stops being believable.
 *
 * @param bytes - A byte count.
 * @returns The size as `900 kB` or `3.4 MB`.
 */
export function formatBytes(bytes: number): string {
  if (bytes < 1000) return `${Math.max(0, Math.round(bytes))} B`
  if (bytes < 1_000_000) return `${Math.round(bytes / 1000)} kB`
  return `${(bytes / 1_000_000).toFixed(1)} MB`
}

/**
 * What the attached files weigh.
 *
 * @param items - The files currently attached.
 * @returns Their total decoded size in bytes.
 */
export function attachmentTotal(items: readonly AgentAttachment[]): number {
  return items.reduce((sum, item) => sum + item.size, 0)
}

/**
 * Why this file cannot be added, or null when it can.
 *
 * Every message names the file and the number it broke, because "too large" on
 * its own leaves the operator guessing which file and by how much.
 *
 * @param items - What is already attached, including files added earlier in
 *   this same drop.
 * @param file - The candidate.
 * @returns The reason, written for the operator, or null.
 */
export function rejectReason(items: readonly AgentAttachment[], file: File): string | null {
  if (items.length >= MAX_ATTACHMENTS) {
    return `A message can carry ${MAX_ATTACHMENTS} files. ${file.name} would be one too many.`
  }
  if (file.size > MAX_ATTACHMENT_BYTES) {
    return `${file.name} is ${formatBytes(file.size)}, over the ${formatBytes(
      MAX_ATTACHMENT_BYTES
    )} limit for one file.`
  }
  if (attachmentTotal(items) + file.size > MAX_TOTAL_BYTES) {
    return `${file.name} would take this message over the ${formatBytes(
      MAX_TOTAL_BYTES
    )} limit for one turn.`
  }
  return null
}

/**
 * The name to show and send.
 *
 * Trimmed to what the server keeps, and stripped of anything path-like, so the
 * label in the composer is the label in the prompt. A browser does not hand out
 * a real path, but a dragged file from an archive tool can carry separators in
 * its name.
 *
 * @param name - The file's own name.
 * @returns A short, path-free label.
 */
function labelOf(name: string): string {
  const bare = name.split(/[\\/]/).pop() || 'file'
  return bare.slice(0, 120)
}

/**
 * Read one file into the shape the composer holds and the request sends.
 *
 * @param file - The picked, dropped or pasted file.
 * @returns The attachment, with its data URL.
 * @throws Error when the file cannot be read at all, which is a disk or
 *   permission failure rather than anything about its content.
 */
export function readAttachment(file: File): Promise<AgentAttachment> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()
    reader.onerror = () => reject(new Error(`${labelOf(file.name)} could not be read.`))
    reader.onload = () => {
      const dataUrl = typeof reader.result === 'string' ? reader.result : ''
      if (!dataUrl) {
        reject(new Error(`${labelOf(file.name)} could not be read.`))
        return
      }
      attachmentSeq += 1
      const encoded = dataUrl.slice(dataUrl.indexOf(',') + 1)
      const mime = (file.type || '').split(';', 1)[0].trim().toLowerCase()
      resolve({
        id: `attachment-${attachmentSeq}`,
        name: labelOf(file.name),
        kind: IMAGE_BASE64_PREFIXES.some((prefix) => encoded.startsWith(prefix)) ? 'image' : 'text',
        mime,
        size: file.size,
        dataUrl,
      })
    }
    reader.readAsDataURL(file)
  })
}

/**
 * One attachment as the request body carries it.
 *
 * @param item - An attached file.
 * @returns Its `{name, mime, data}` entry, with `mime` omitted where the server
 *   has no rule for it. See the module docstring.
 */
export function attachmentPayload(item: AgentAttachment): AttachmentPayload {
  const payload: AttachmentPayload = { name: item.name, data: item.dataUrl }
  if (DECLARABLE.has(item.mime)) payload.mime = item.mime
  return payload
}

/**
 * One attachment as the sent message remembers it.
 *
 * @param item - An attached file.
 * @returns Its metadata, without the bytes.
 */
export function attachmentMeta(item: AgentAttachment): AgentAttachmentMeta {
  return { name: item.name, kind: item.kind, mime: item.mime, size: item.size }
}

/**
 * The files carried by a drop, a paste or a file input.
 *
 * A `DataTransfer` reports files two ways and neither is reliable alone: Chrome
 * fills `items` for a pasted screenshot and `files` for a dragged one, and a
 * directory drag produces an item with no file behind it at all.
 *
 * @param transfer - The event's `dataTransfer` or `clipboardData`.
 * @returns Every real file it carries, in order, with duplicates removed.
 */
export function filesFrom(transfer: DataTransfer | null | undefined): File[] {
  if (!transfer) return []
  const found: File[] = []
  // Array.from, not indexing: a FileList and a DataTransferItemList are
  // array-like rather than arrays, and neither has the methods of one.
  for (const file of Array.from<File>(transfer.files ?? [])) {
    if (file) found.push(file)
  }
  if (found.length > 0) return found
  for (const entry of Array.from<DataTransferItem>(transfer.items ?? [])) {
    if (entry?.kind !== 'file') continue
    const file = entry.getAsFile()
    if (file) found.push(file)
  }
  return found
}
